File: Java/Eta/Applications/PerfTools/src/main/java/com/thomsonreuters/upa/perftools/upajtransportperf/SessionHandler.java
package com.thomsonreuters.upa.perftools.upajtransportperf;

import java.util.ArrayList;
import java.util.concurrent.locks.ReentrantLock;

import com.thomsonreuters.upa.perftools.common.TimeRecordQueue;
import com.thomsonreuters.upa.transport.Channel;
import com.thomsonreuters.upa.transport.MCastStats;
import com.thomsonreuters.upa.transport.TransportFactory;

/** High level handler for multiple transport session threads.
  * Enables a transport thread to handle multiple channels. */
public class SessionHandler
{
	private ReentrantLock			_handlerLock;		/* Lock for the handler. */
	private ArrayList<Channel>		_newChannelsList;	/* New channels from the main thread are passed to the handler through this queue. */
	private int						_openChannelsCount;	/* Total number of channels this thread is currently handling. */
	private boolean					_active;			/* Connections are active, run the test */
	private TimeRecordQueue			_latencyRecords;	/* Latency records from the TransportThread. */
	private TransportThread			_transportThread;	/* Thread associated with this handler. */
	private int						_role;				/* Role of this handler. */
	private MCastStats				_prevMCastStats;	/* Stores any multicast statistics. */
	
	/**
	 * Initializes the handler for a transport perf thread.
	 */
	public void init()
	{
		_handlerLock = new ReentrantLock();
		_newChannelsList = new ArrayList<Channel>(10000);
		_active = false;
		_latencyRecords = new TimeRecordQueue();

		_role = TransportTestRole.READER | TransportTestRole.WRITER;
		_prevMCastStats = TransportFactory.createMCastStats();
	}
	
	/**
	 * Handler lock.
	 *
	 * @return Lock for the handler.
	 */
	public ReentrantLock handlerLock()
	{
		return _handlerLock;
	}
	
	/**
	 * New channels list.
	 *
	 * @return New channels accepted from the main thread.
	 */
	public ArrayList<Channel> newChannelsList()
	{
		return _newChannelsList;
	}
	
	/**
	 * Open channels count.
	 *
	 * @return Total number of channels this thread is currently handling.
	 */
	public int openChannelsCount()
	{
		return _openChannelsCount;
	}

	/**
	 * Set the total number of channels this thread is currently handling. 
	 *
	 * @param openChannelsCount the open channels count
	 */
	public void openChannelsCount(int openChannelsCount)
	{
		_openChannelsCount = openChannelsCount;
	}

	/**
	 * Active.
	 *
	 * @return true if connections are active.
	 */
	public boolean active()
	{
		return _active;
	}

	/**
	 * Active.
	 *
	 * @param active - connections are active.
	 */
	public void active(boolean active)
	{
		_active = active;
	}

	/**
	 * Latency records.
	 *
	 * @return Latency records. Updated from transport thread.
	 */
	public TimeRecordQueue latencyRecords()
	{
		return _latencyRecords;
	}

	/**
	 * Transport thread.
	 *
	 * @return Thread associated with this handler.
	 */
	public TransportThread transportThread()
	{
		return _transportThread;
	}

	/**
	 * 
	 * @param transportThread Thread associated with this handler.
	 */
	void transportThread(TransportThread transportThread)
	{
		_transportThread = transportThread;
	}

	/**
	 * Role.
	 *
	 * @return Role of this handler.
	 */
	public int role()
	{
		return _role;
	}

	/**
	 * Role.
	 *
	 * @param role  Role of this handler.
	 */
	public void role(int role)
	{
		_role = role;
	}

	/**
	 * Prev M cast stats.
	 *
	 * @return multicast statistics. Used for multicast connection.
	 */
	public MCastStats prevMCastStats()
	{
		return _prevMCastStats;
	}

	/**
	 * Prev M cast stats.
	 *
	 * @param prevMCastStats multicast statistics.
	 */
	public void prevMCastStats(MCastStats prevMCastStats)
	{
		_prevMCastStats = prevMCastStats;
	}
}

File: CSharp/Eta/Applications/PerfTools/TransportPerf/SessionHandler.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2023-2024 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

using System.Collections.Generic;
using System.Threading;
using LSEG.Eta.PerfTools.Common;
using LSEG.Eta.Transports;

namespace LSEG.Eta.PerfTools.TransportPerf
{
	/// <summary>
	/// High level handler for multiple transport session threads.
	/// Enables a transport thread to handle multiple channels.
	/// </summary>
	public class SessionHandler
    {
		/// <summary>
		/// Gets Handler lock.
		/// </summary>
		public ReaderWriterLockSlim HandlerLock { get; private set; } =
			new ReaderWriterLockSlim(LockRecursionPolicy.SupportsRecursion);

		/// <summary>
		/// Gets new channels accepted from the main thread.
		/// </summary>
		public List<IChannel> NewChannelsList { get; private set; } = new List<IChannel>(10000);

		/// <summary>
		/// Gets or sets total number of channels this thread is currently handling.
		/// </summary>
		public int OpenChannelsCount { get; set; }

		/// <summary>
		/// Gets or sets whether connections are active.
		/// </summary>
		public bool Active { get; set; }

		/// <summary>
		/// Gets latency records. Updated from transport thread.
		/// </summary>
		public TimeRecordQueue LatencyRecords { get; private set; } = new TimeRecordQueue();

		/// <summary>
		/// Gets or sets a thread associated with this handler.
		/// </summary>
		public TransportThread? TransportThread { get; set; }
		
		/// <summary>
		/// Gets or sets a role of this handler.
		/// </summary>
		public TransportTestRole Role { get; set; }

		/// <summary>
		/// Initializes the handler for a transport perf thread.
		/// </summary>
		public void Init()
		{
			Active = false;
			Role = TransportTestRole.READER | TransportTestRole.WRITER;
		}
	}
}
